pkg/apis/pipeline/v1beta1/pipeline_validation_test.go

```diff
@@ -106,653 +62,948 @@ func TestPipeline_Validate(t *testing.T) {
 		p: &v1beta1.Pipeline{
 			ObjectMeta: metav1.ObjectMeta{Name: "pipeline"},
 			Spec: v1beta1.PipelineSpec{
-				Tasks: []v1beta1.PipelineTask{
-					{Name: "foo"},
-				},
+				Description: "this is a valid pipeline with all possible fields initialized",
+				Resources: []v1beta1.PipelineDeclaredResource{{
+					Name:     "app-repo",
+					Type:     "git",
+					Optional: false,
+				}, {
+					Name:     "app-image",
+					Type:     "git",
+					Optional: false,
+				}},
+				Tasks: []v1beta1.PipelineTask{{
+					Name:    "my-task",
+					TaskRef: &v1beta1.TaskRef{Name: "foo-task"},
+					Retries: 5,
+					Resources: &v1beta1.PipelineTaskResources{
+						Inputs: []v1beta1.PipelineTaskInputResource{{
+							Name:     "task-app-repo",
+							Resource: "app-repo",
+						}},
+						Outputs: []v1beta1.PipelineTaskOutputResource{{
+							Name:     "task-app-image",
+							Resource: "app-image",
+						}},
+					},
+					Params: []v1beta1.Param{{
+						Name:  "param1",
+						Value: v1beta1.ArrayOrString{},
+					}},
+					Workspaces: []v1beta1.WorkspacePipelineTaskBinding{{
+						Name:      "task-shared-workspace",
+						Workspace: "shared-workspace",
+					}},
+					Timeout: &metav1.Duration{Duration: 5 * time.Minute},
+				}},
+				Params: []v1beta1.ParamSpec{{
+					Name:        "param1",
+					Type:        v1beta1.ParamType("string"),
+					Description: "this is my param",
+					Default: &v1beta1.ArrayOrString{
+						Type:      v1beta1.ParamType("string"),
+						StringVal: "pipeline-default",
+					},
+				}},
+				Workspaces: []v1beta1.WorkspacePipelineDeclaration{{
+					Name:        "shared-workspace",
+					Description: "this is my shared workspace",
+				}},
+				Results: []v1beta1.PipelineResult{{
+					Name:        "pipeline-result",
+					Description: "this is my pipeline result",
+					Value:       "pipeline-result-default",
+				}},
 			},
 		},
-		failureExpected: true,
-	}, {
-		name: "pipeline spec with taskref and taskspec",
-		p: &v1beta1.Pipeline{
-			ObjectMeta: metav1.ObjectMeta{Name: "pipeline"},
-			Spec: v1beta1.PipelineSpec{
-				Tasks: []v1beta1.PipelineTask{
-					{
-						Name:    "foo",
-						TaskRef: &v1beta1.TaskRef{Name: "foo-task"},
-						TaskSpec: &v1beta1.TaskSpec{
-							Steps: []v1beta1.Step{{
-								Container: corev1.Container{Name: "foo", Image: "bar"},
-							}},
-						},
-					},
+	}}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := tt.p.Validate(context.Background())
+			if (!tt.failureExpected) && (err != nil) {
+				t.Errorf("Pipeline.Validate() returned error: %v", err)
+			}
+
+			if tt.failureExpected && (err == nil) {
+				t.Error("Pipeline.Validate() did not return error, wanted error")
+			}
+		})
+	}
+}
+
+func TestPipelineSpec_Validate(t *testing.T) {
+	tests := []struct {
+		name            string
+		ps              *v1beta1.PipelineSpec
+		failureExpected bool
```

Comment:
this is definitely outside the scope of your changes b/c the tests were already here but I really recommend having completely different tests for failure/non-failure cases in general - makes them much easier to read and maintain!

e.g. if you have a test where all the pipelines are invalid, and one where they are all valid, makes it much easier to understand whats covered and whats not

Reply:
in fact it looks at a glance like all the cases here might be invalid so maybe the flag isnt needed?

Reply:
yup, will create separate functions for failure tests vs non-failure tests.